### ISSUE TITLE
Abstract_Class_Parser.hh uses deprecated C++ feature pointer_to_unary_function

### ISSUE THREAD

**Issue by KineticTheory** [lanl/Draco#775]
It looks like `Abstract_Class_Parser.hh` uses ` std::pointer_to_unary_function` which is [deprecated and will be removed in C++17.](https://en.cppreference.com/w/cpp/utility/functional/pointer_to_unary_function)  Newer compilers like LLVM version 9+ complain about it:

```console
Building CXX object src/parser/test/CMakeFiles/Ut_parser_tstAbstract_Class_Parser_exe.dir/tstAbstract_Class_Parser.cc.o
cd /scratch/kellyt/llvm-mpid/d/src/parser/test && /scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/gcc-9.2.0/llvm-9.0.0-tlfcu5ryd2komz3xgoh6yidg5lkg2i5w/bin/clang++  -DPROJECT_BINARY_DIR=\"/scratch/kellyt/llvm-mpid/d/src/parser/test\" -DPROJECT_SOURCE_DIR=\"/home/kellyt/draco/src/parser/test\" -D_POSIX_C_SOURCE=200112 -D_XOPEN_SOURCE=600 -D__STDC_CONSTANT_MACROS -I/scratch/kellyt/llvm-mpid/d/src/c4 -I/home/kellyt/draco/src -I/scratch/kellyt/llvm-mpid/d/src/ds++ -isystem /scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/clang-9.0.0/libquo-1.3-5qm7zvlbuali4keevyb6db7zzw5d4nsl/include -isystem /scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/clang-9.0.0/openmpi-3.1.4-e3ziqwg77khtelehq73ijivqetk3rjqh/include -isystem /scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/clang-9.0.0/openmpi-3.1.4-e3ziqwg77khtelehq73ijivqetk3rjqh/lib  -Wcast-align -Wpointer-arith -Wall -Wno-long-long -Wreserved-id-macro -pedantic -Wno-undefined-var-template -Wno-potentially-evaluated-expression -stdlib=libc++ -fopenmp=libomp   -Werror  -g -fno-inline -O0 -Wextra -DDEBUG -Woverloaded-virtual -fPIE   -fexceptions -pthread -std=c++14 -o CMakeFiles/Ut_parser_tstAbstract_Class_Parser_exe.dir/tstAbstract_Class_Parser.cc.o -c /home/kellyt/draco/src/parser/test/tstAbstract_Class_Parser.cc
In file included from /home/kellyt/draco/src/parser/test/tstAbstract_Class_Parser.cc:13:
/home/kellyt/draco/src/parser/Abstract_Class_Parser.hh:93:37: error: 'pointer_to_unary_function<rtt_parser::Token_Stream &, std::__1::shared_ptr<Parent> >' is deprecated [-Werror,-Wdeprecated-declarations]
          typename Parse_Function = pointer_to_unary_function<
                                    ^
/home/kellyt/draco/src/parser/test/tstAbstract_Class_Parser.cc:71:5: note: in instantiation of default argument for 'Abstract_Class_Parser<Parent, &rtt_parser::Class_Parse_Table<Parent>::get_parse_table_, &rtt_parser::Class_Parse_Table<Parent>::get_parsed_object_>' required here
    Abstract_Class_Parser<Parent, get_parse_table_,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/gcc-9.2.0/llvm-9.0.0-tlfcu5ryd2komz3xgoh6yidg5lkg2i5w/bin/../include/c++/v1/functional:1088:28: note: 'pointer_to_unary_function<rtt_parser::Token_Stream &, std::__1::shared_ptr<Parent> >' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX11 pointer_to_unary_function
                           ^
/scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/gcc-9.2.0/llvm-9.0.0-tlfcu5ryd2komz3xgoh6yidg5lkg2i5w/bin/../include/c++/v1/__config:944:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX11'
#  define _LIBCPP_DEPRECATED_IN_CXX11 _LIBCPP_DEPRECATED
                                      ^
/scratch/vendors/spack.20191108/opt/spack/linux-rhel7-sandybridge/gcc-9.2.0/llvm-9.0.0-tlfcu5ryd2komz3xgoh6yidg5lkg2i5w/bin/../include/c++/v1/__config:933:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
1 error generated.
make[2]: *** [src/parser/test/CMakeFiles/Ut_parser_tstAbstract_Class_Parser_exe.dir/tstAbstract_Class_Parser.cc.o] Error 1
```

This blocks our move to C++17 and needs to be updated.

**Comment by KineticTheory** [comment]
cc @jhchang-lanl @attom @alexrlongne 

This issue can be reproduced on ccs-net machines with this environment and a regular configure & build of draco:

```console
module load user_contrib/2019.11
module load numdiff python random123 llvm cmake openmpi 
```

**Comment by KineticTheory** [comment]
Fixed by #786
